packages/core/src/utils/collectHeadings.ts

```diff
@@ -0,0 +1,67 @@
+// ToC: get the html nodelist for headings
+interface TocSection {
+  id: string;
+  title: string;
+  level: string;
+  children?: any;
+}
+
+export function collectHeadings(nodes: NodeListOf<HTMLHeadingElement>) {
+  const sections: Array<TocSection> = [];
+  const headerTitle = document.querySelector("main header h1");
+
+  Array.from(nodes).forEach((node) => {
+    const pageHeading: boolean = node.parentElement?.firstChild === node;
+    const { id, innerText: title, tagName: level } = node;
+
+    if (!(id && title)) {
+      return;
+    }
+
+    if (level === "H1" && (headerTitle ?? !pageHeading)) {
+      sections.push({ id, title, level, children: [] });
+    }
+
```

Comment:
Can you please explain what you're trying to achieve here? Especially on line 21.

Reply:
In the issue:
> I would assume h1 other than very first "title" h1 would show up in ToC 

Here I'm checking if there's a title in the frontmatter. If there is one, then don't remove the h1 heading from ToC  eg. if markdown starts with `# Some Heading`. If no frontmatter title is set, then h1 is considered as page's title and we remove the h1. 

The only caveat with this approach is that the frontmatter title should be inside a header element (that's how our docs and blog layouts are set up right now).

Reply:
Also as in the shared image above
> even if h1 does not show up i would imagine the h2/h3 etc underneath it would show up ...

Maybe I understood this wrong. What I tried here was to show h2/h3 even if they didn't have a parent. Let me know if this is not the desired outcome and I'll change that :)

Reply:
> They show up as seen above right ?
But I don't think that's a correct representation of the page structure. Take a look at the screenshot once again. You have "Some heading lvl 2" (h2 tag) with the same indentation as "User experience" (h1 tag) 😛

Reply:
> I didn't understand the 2nd option, aren't 2 and 3 the same thing ?

Only for pages with multiple h1 tags.

Reply:
> Yeah, I think option 3 has better maintainability. I will make that change.

I think so too.

Reply:
But now I'm facing another issue with the above ie., what happens in pages where there are only h2 headings for example like we have in our /docs page. In this scenario, the ToC won't render.

Looking at tailwind's [syntax](https://tailwindui.com/templates/syntax) theme, they are only displaying h2 and h3. if we put an h3 without an h2 before it, an error is thrown like below.

<img width="1182" alt="Screen Shot 2023-01-23 at 10 15 58 PM" src="https://user-images.githubusercontent.com/42637597/214129377-d845a0e4-49a8-43a8-9ec5-317608daaf24.png">

Should we not render h2 if there's no preceding h1 and same for h3  ?

Reply:
@khalilcodes I think let's just not remove anything from the ToC. Even we have some pages with multiple h1 tags and I used to write my notes using multiple h1 tags as well. I have also googled around a bit, and while this may be a good idea to only have one h1, there's nothing wrong with having multiple. It all depends on what content you're writing and probably also how many levels of different sections you need, and you may need many, so you'd probably could also use this h1 to additionaly divide your content. The other benefit of not removing h1 tags is that if they do have some text, like e.g. intro to the page, if you remove the first h1 you can't use the right hand side ToC to quickly jump to the very first heading.

Reply:
And in this case the ToC clearly reflects all the different sections on your page, nothing less nothing more. Probably even for screen readers this might be a better approach.